### PR TITLE
Kconfig: add CONFIG_BOARD_REVISION

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
Propagate the board revision to Kconfig via the environment.

An alternative to: https://github.com/zephyrproject-rtos/zephyr/pull/44131

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
To try it out, do something like this:

```
$ west build --cmake-only -p -b nrf9160dk_nrf9160@0.7.0 samples/hello_world/
$ grep BOARD_REVISION build/zephyr/.config
CONFIG_BOARD_REVISION="0.7.0"
```

File: arch/Kconfig [hunk 1]
```diff
@@ -846,6 +846,9 @@ config BOARD
 	  if not found we look for the linker file in
 	  soc/<arch>/<family>/<series>
 
+config BOARD_REVISION
+	def_string "$BOARD_REVISION"
```
Comment (stephanosio) [inline arch/Kconfig:850]: would be nice to add `help` as well.

by the way, is `arch/Kconfig` really the right place to add this config? IMO, `boards/Kconfig` would be more suited for this purpose.

**Review by stephanosio — COMMENTED** [review]
If we are going to add this as a Kconfig symbol, this would indeed be the correct thing to do.

No strong opinion on whether this should be a Kconfig symbol or just a good old `#define BOARD_REVISION`; though, I am inclined to say Kconfig is better since you can potentially reuse this to set the defaults for other configs.

**Review by carlescufi — DISMISSED** [review]
I like this better.

**Review by jfischer-no — DISMISSED** [review]

File: arch/Kconfig (hunk 1 above)
Comment (carlescufi) [reply arch/Kconfig:850]: Agreed, `arch/Kconfig` is not the right place for this. +1 for `boards/Kconfig`.

**Review by nordicjm — DISMISSED** [review]
Much simpler and better!

**Review by nashif — CHANGES_REQUESTED** [review]
what @stephanosio  says...

File: arch/Kconfig (hunk 1 above)
Comment (mbolivar-nordic) [reply arch/Kconfig:850]: This is being added right under the definition for CONFIG_BOARD. Are you saying we should move both of them to boards/? Moving just this one would not make sense. 

@nashif @stephanosio

**Comment by mbolivar-nordic** [comment]
@nashif I have addressed @stephanosio's comments, including adding a prep work patch moving CONFIG_BOARD to boards/Kconfig to keep everything together.

**Review by nashif — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

File: boards/Kconfig [hunk 2]
```diff
@@ -9,6 +9,14 @@ config BOARD
 	  if not found we look for the linker file in
 	  soc/<arch>/<family>/<series>
 
+config BOARD_REVISION
+	def_string "$BOARD_REVISION"
```
Comment (tejlmand) [inline boards/Kconfig:13]: @mbolivar-nordic a little late to the game.

I'm sure this is useful but did you notice the fact that the `BOARD` Kconfig setting is not passed through the environment, but set in the board's Kconfig.

The result of that is that the BOARD value in Kconfig != BOARD value passed to CMake, se for example:
https://github.com/zephyrproject-rtos/zephyr/blob/a57dee94d68f2aa0f1a6a07512f1ceab6248c982/boards/arm/nrf5340dk_nrf5340/Kconfig.defconfig#L8-L9

This means that user can fetch the revision for the board given with `-DBOARD=<board>@<rev>`, but the board available in Kconfig may not be the same board, for example:
`-DBOARD=nrf5340dk_nrf5340_cpuapp_ns@1.2.3` results in:
- `CONFIG_BOARD == nrf5340dk_nrf5340_cpuapp`
- `CONFIG_BOARD_REVISION == 1.2.3`

so not sure passing revision this way is wise if we don't cleanup and enforce that `CONFIG_BOARD` is identical to the board provided by the user with `-DBOARD`.

**Review by tejlmand — COMMENTED** [review]
a late arrival comment.
FYI @stephanosio @nashif @carlescufi @mbolivar-nordic

File: boards/Kconfig (hunk 2 above)
Comment (mbolivar-nordic) [reply boards/Kconfig:13]: > so not sure passing revision this way is wise if we don't cleanup and enforce that `CONFIG_BOARD` is identical to the board provided by the user with `-DBOARD`.

are you aware of any counterexamples? That would indeed seem like a bug in the board defconfig if CONFIG_BOARD ended up different than BOARD (without the revision), wouldn't you agree?

Comment (tejlmand) [reply boards/Kconfig:13]: > are you aware of any counterexamples? That would indeed seem like a bug in the board defconfig if CONFIG_BOARD ended up different than BOARD (without the revision), wouldn't you agree?

@mbolivar-nordic did you look at the code I linked ?

That is exactly the counter example you are requesting.

A user specifying `-DBOARD=nrf5340dk_nrf5340_cpuapp_ns@0.1.2` will have the BOARD = nrf5340dk_nrf5340_cpuapp_ns is CMake, but as BOARD = nrf5340dk_nrf5340_cpuapp in Kconfig, that is without `_ns` in Kconfig.

**Pull request merged**